cryptography/src/bls12381/dkg.rs

```diff
@@ -0,0 +1,1702 @@
+        self.summary.write(buf);
+        self.players.write(buf);
+        self.public.write(buf);
+    }
+}
+
+impl<V: Variant, P: PublicKey> Read for Output<V, P> {
+    type Cfg = NonZeroU32;
+
+    fn read_cfg(
+        buf: &mut impl bytes::Buf,
+        &max_players: &Self::Cfg,
+    ) -> Result<Self, commonware_codec::Error> {
+        Ok(Self {
+            summary: ReadExt::read(buf)?,
+            players: Read::read_cfg(buf, &(RangeCfg::new(1..=max_players.get() as usize), ()))?,
+            public: Read::read_cfg(buf, &RangeCfg::from(NZU32!(1)..=max_players))?,
+        })
+    }
+}
+
+/// Information about the current round of the DKG.
+///
+/// This is used to bind signatures to the current round, and to provide the
+/// information that dealers, players, and observers need to perform their actions.
+#[derive(Debug, Clone)]
+pub struct Info<V: Variant, P: PublicKey> {
+    round: u64,
+    previous: Option<Output<V, P>>,
+    dealers: Ordered<P>,
+    players: Ordered<P>,
+    /// Never written when encoded, always computed from the previous fields.
+    summary: Summary,
+}
+
+impl<V: Variant, P: PublicKey> PartialEq for Info<V, P> {
+    fn eq(&self, other: &Self) -> bool {
+        self.summary == other.summary
+    }
+}
+
+impl<V: Variant, P: PublicKey> Info<V, P> {
+    /// Figure out what the dealer share should be.
+    ///
+    /// If there's no previous round, we need a random value, hence `rng`.
+    ///
+    /// However, if there is a previous round, we expect a share, hence `Result`.
+    fn generate_dealer_share_if_necessary(
+        &self,
+        mut rng: impl CryptoRngCore,
+        share: Option<Scalar>,
+    ) -> Result<Scalar, Error> {
+        let out = match (self.previous.as_ref(), share) {
+            (None, None) => Scalar::from_rand(&mut rng),
+            (_, Some(x)) => x,
+            (Some(_), None) => return Err(Error::MissingDealerShare),
+        };
+        Ok(out)
+    }
+
+    fn degree(&self) -> u32 {
+        quorum(self.players.len() as u32).saturating_sub(1)
+    }
+
+    fn threshold(&self) -> u32 {
+        self.degree() + 1
+    }
+
+    fn required_commitments(&self) -> u32 {
+        let dealer_quorum = quorum(self.dealers.len() as u32);
+        let prev_quorum = self
+            .previous
+            .as_ref()
+            .map(Output::quorum)
+            .unwrap_or(u32::MIN);
+        dealer_quorum.max(prev_quorum)
+    }
+
+    fn max_reveals(&self) -> u32 {
+        max_faults(self.players.len() as u32)
+    }
+
+    fn player_index(&self, player: &P) -> Result<u32, Error> {
+        self.players
+            .position(player)
+            .map(|x| x as u32)
+            .ok_or(Error::UnknownPlayer)
+    }
+
+    fn dealer_index(&self, dealer: &P) -> Result<u32, Error> {
+        self.dealers
+            .position(dealer)
+            .map(|x| x as u32)
+            .ok_or(Error::UnknownDealer(format!("{dealer:?}")))
+    }
+
+    #[must_use]
```

Comment:
Dumb question: I've seen these defined on structs only before...this means that if we have `Info`, we'll see a compiler warning if both of these haven't been used?